docs/design_notes/lift.md

```diff
@@ -0,0 +1,338 @@
+# Lifted transformation
+
+⚠️ Advanced topic ⚠️
+
+This design note explains the underlying implementation of `flax.linen.transform` which enable JAX transformations inside `Module`.
+
+
+## Introduction
+
+JAX uses a functional API meaning that it only garantuees correct behavior when using functions without side effects.
+Typically these side effects are the result of mutating an object that lives outside of the function.
+
+The functional paradigm has some advantages like the ability to reason about state and stochasticity explicitly.
+A function output can only change when the input arguments change.
+Therefore, a function is garantueed to behave deterministicly.
+
+But pure functions offer another big advantage to JAX specifically: They enable functional transformations.
+For example `jax.vmap(f)` will vectorize a function `f`.
+Because `f` cannot have side-effects the vectorized/parallel version of `f` is well-defined. To see why we need this restriction, consider what happens if `f` would increment a counter or draw a random number.
```

Comment:
Suggestion: "side effects" without the dash (for consistency)

```suggestion
Because `f` cannot have side effects the vectorized/parallel version of `f` is well-defined. To see why we need this restriction, consider what happens if `f` would increment a counter or draw a random number.
```